Mathlib/Dynamics/TopologicalEntropy/CoverEntropy.lean

```diff
@@ -0,0 +1,590 @@
+    exact le_top
+  let u := fun _ : ℕ ↦ log (coverMincard T F U n) / n
+  let v := fun m : ℕ ↦ log (coverMincard T F U n) / m
+  let w := fun m : ℕ ↦ log (coverMincard T F (U ○ U) m) / m
+  have key : w ≤ᶠ[atTop] u + v :=
+    eventually_atTop.2 ⟨1, fun m m_pos ↦ log_coverMincard_comp_le F_inv U_symm n_pos m_pos⟩
+  apply ((limsup_le_limsup) key).trans
+  suffices h : atTop.limsup v = 0 by
+    have := @limsup_add_le_add_limsup ℕ atTop u v
+    rw [h, add_zero] at this
+    specialize this (Or.inr EReal.zero_ne_top) (Or.inr EReal.zero_ne_bot)
+    exact le_of_le_of_eq this (limsup_const (log (coverMincard T F U n) / n))
+  exact Tendsto.limsup_eq (EReal.tendsto_const_div_atTop_nhds_zero_nat logm_nneg logm_fin)
+
+lemma coverEntropySupUni_le_coverEntropyInfUni {T : X → X} {F : Set X} (F_inv : MapsTo T F F)
+    {U : Set (X × X)} (U_symm : SymmetricRel U) :
+    coverEntropySupUni T F (U ○ U) ≤ coverEntropyInfUni T F U :=
+  (le_liminf_of_le) (eventually_atTop.2
+  ⟨1, fun m m_pos ↦ coverEntropySupUni_le_log_coverMincard_div F_inv U_symm m_pos⟩)
+
+lemma coverEntropyInfUni_finite_of_isCompact_invariant [UniformSpace X] {T : X → X} {F : Set X}
+    (F_comp : IsCompact F) (F_inv : MapsTo T F F) {U : Set (X × X)} (U_uni : U ∈ 𝓤 X) :
+    coverEntropyInfUni T F U < ⊤ := by
+  rcases comp_symm_mem_uniformity_sets U_uni with ⟨V, V_uni, V_symm, V_U⟩
+  rcases exists_isDynCoverOf_of_isCompact_invariant F_comp F_inv V_uni 1 with ⟨s, s_cover⟩
+  apply lt_of_le_of_lt (coverEntropyInfUni_antitone T F V_U)
+  apply lt_of_le_of_lt (coverEntropyInfUni_le_log_card_div F_inv V_symm zero_lt_one s_cover)
+  rw [Nat.cast_one, div_one, log_lt_top_iff, ← ENat.toENNReal_top]
+  exact_mod_cast Ne.lt_top (ENat.coe_ne_top (Finset.card s))
+
+lemma coverEntropySupUni_finite_of_isCompact_invariant [UniformSpace X] {T : X → X} {F : Set X}
+    (F_comp : IsCompact F) (F_inv : MapsTo T F F) {U : Set (X × X)} (U_uni : U ∈ 𝓤 X) :
+    coverEntropySupUni T F U < ⊤ := by
+  rcases comp_symm_mem_uniformity_sets U_uni with ⟨V, V_uni, V_symm, V_U⟩
+  exact lt_of_le_of_lt (coverEntropySupUni_antitone T F V_U)
+    <| lt_of_le_of_lt (coverEntropySupUni_le_coverEntropyInfUni F_inv V_symm)
+    <| coverEntropyInfUni_finite_of_isCompact_invariant F_comp F_inv V_uni
+
+/-! ### Cover entropy -/
+
+/-- The entropy of `T` restricted to `F`, obtained by taking the supremum over entourages.
+  Note that this supremum is approached by taking small uniformities. This first version uses a
+  `liminf`.-/
+noncomputable def coverEntropyInf [UniformSpace X] (T : X → X) (F : Set X) :=
+  ⨆ U ∈ 𝓤 X, coverEntropyInfUni T F U
+
+/-- The entropy of `T` restricted to `F`, obtained by taking the supremum over entourages.
+  Note that this supremum is approached by taking small uniformities. This second version uses a
+  `limsup`.-/
+noncomputable def coverEntropySup [UniformSpace X] (T : X → X) (F : Set X) :=
+  ⨆ U ∈ 𝓤 X, coverEntropySupUni T F U
```

Comment:
I have a problem that, with the way the results are currently phrased, the end user will not know if he should use `coverEntropInf` or `coverEntropySup` as the canonical one (since they coincide in the most relevant cases) and different users might pick different choices. I guess it's worth making one choice more canonical than the other by using the name `coverEntropy` -- and explaining carefully in the docstrings, both of the definition and the file-level docstring, what is going on here.

Reply:
Good idea.

I've chosen to make the `limsup` definition canonical. I don't think it changes much, but there is a remote reason why it might avoid some rewrites (thanks to `coverEntropyEnt_le_log_card_div`, getting upper bounds on the `liminf`/`limsup` versions is essentially the same ; lower bounds on the entropy are more tricky, and easier with a `limsup` than a `liminf`).